### PR TITLE
Int 409 additional information required if screen_out and evaluate_out are selected

### PR DESCRIPTION
**Pull request by NagaPradeepGajula-osi**
### Jira Story

- [Additional Information field is required when Decision is `Screen out` AND Category is `Evaluate Out` INT-409](https://osi-cwds.atlassian.net/browse/INT-409)

### Purpose
additional information includes required if decision value is selected as screen_out and if decision_detail(category) value is selected as evaluate_out, throughs an error if additional information is left empty.

### CODE REVIEW

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js [hunk 1]
```diff
@@ -153,6 +163,37 @@ describe('allegationShowSelectors', () => {
...
+
+      it('does not include an error message if decision is screen out, and decision detail is empty', () => {
+        const screening = {screening_decision: 'screen_out'}
+        const state = fromJS({screening})
+        expect(getErrorsSelector(state).get('additional_information'))
+          .toEqualImmutable(List())
+      })
+    })
+
+    describe('getAdditionalInfoRequiredSelector', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: Scope is messed up. `getAdditionalInfoRequiredSelector` should not be wrapped under `getErrorsSelector` block.

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js [hunk 2]
```diff
@@ -24,6 +26,14 @@ describe('allegationShowSelectors', () => {
     })
   })
 
+  describe('getAdditionalInformationSelector', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: Selector also returns `value` which is not covered in tests.

File: app/javascript/selectors/screening/decisionShowSelectors.js [hunk 3]
```diff
@@ -4,14 +4,16 @@ import {getScreeningSelector} from 'selectors/screeningSelectors'
 import SCREENING_DECISION from 'enums/ScreeningDecision'
 import SCREENING_DECISION_OPTIONS from 'enums/ScreeningDecisionOptions'
 import {isRequiredCreate, isRequiredIfCreate, combineCompact} from 'utils/validator'
+import {getDecisionValueSelector, getDecisionDetailValueSelector} from './decisionFormSelectors'
```
Comment (AmanShurafa) [inline app/javascript/selectors/screening/decisionShowSelectors.js]: Form & Show use different parts of the store. You should not use Form selectors in Show Selectors.

File: app/javascript/selectors/screening/decisionShowSelectors.js [hunk 4]
```diff
@@ -33,13 +35,24 @@ export const getErrorsSelector = createSelector(
...
       restrictions_rationale: combineCompact(
         isRequiredIfCreate(restrictionsRationale, 'Please enter an access restriction reason', () => (accessRestrictions))
       ),
     })
   )
 )
 
+export const getAdditionalInfoRequiredSelector = createSelector(
+  getDecisionValueSelector,
+  getDecisionDetailValueSelector,
```
Comment (AmanShurafa) [inline app/javascript/selectors/screening/decisionShowSelectors.js]: Don't use Form selectors. Decision Form selectors select data from `screeningDecisionForm` while Show selectors select data from `screening`. Use:
```
(state) => state.getIn(['screening', 'screening_decision']),
(state) => state.getIn(['screening', 'screening_decision_detail']),
```

Or better yet, create `getDecisionValueSelector` and `getDecisionDetailValueSelector` selectors since their being used more than once and reduce the complexity of `getErrorsSelector`.

File: app/javascript/views/ScreeningDecisionForm.jsx [hunk 5]
```diff
@@ -65,12 +66,20 @@ const ScreeningDecisionForm = ({
...
+            required = {isAdditionalInfoRequired}
+          >
+            <textarea
+              id='additional_information'
+              onChange={({target: {value}}) => onChange('additional_information', value)}
+              value={additionalInformation.value}
+              onBlur={() => onBlur('additional_information')}
+              maxLength='255'
+            />
+          </FormField>
```
Comment (AmanShurafa) [inline app/javascript/views/ScreeningDecisionForm.jsx]: TextField component was created 2 months ago but never merged 😢:
https://github.com/ca-cwds/intake/commit/9ef009f36accd99ecc8851804138f9cbe4634fd7

File: spec/javascripts/views/ScreeningDecisionShowSpec.jsx [hunk 6]
```diff
@@ -92,6 +92,13 @@ describe('ScreeningDecisionShow', () => {
     expect(additionalInformation.children().text()).toEqual('My additional information')
   })
 
+  it('renders the additional information label with required', () => {
+    const component = renderScreeningDecisionShow({isAdditionalInfoRequired: true})
+    const additionalInformationLabel = component.find('ShowField[label="Additional information"]')
+    expect(additionalInformationLabel.exists()).toEqual(true)
```
Comment (AmanShurafa) [inline spec/javascripts/views/ScreeningDecisionShowSpec.jsx]: This line is not needed since the next line covers it. If label does not exist, the line below will blow up. "Cannot call props() on undefined" or something like that.

File: spec/javascripts/views/ScreeningDecisionShowSpec.jsx [hunk 7]
```diff
@@ -92,6 +92,13 @@ describe('ScreeningDecisionShow', () => {
     expect(additionalInformation.children().text()).toEqual('My additional information')
   })
 
+  it('renders the additional information label with required', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/views/ScreeningDecisionShowSpec.jsx]: Need another test for when field is not required.

File: spec/javascripts/views/ScreeningDecisionFormSpec.jsx [hunk 8]
```diff
@@ -273,12 +275,19 @@ describe('ScreeningDecisionForm', () => {
...
     it('renders the additional information text area with the proper default props', () => {
       const component = renderScreeningDecisionForm({})
-      const additionalInformationLabel = component.find('label[htmlFor="additional_information"]')
+      const additionalInformationLabel = component.find('FormField[htmlFor="additional_information"]')
       expect(additionalInformationLabel.exists()).toEqual(true)
       const additionalInformationTextArea = component.find('textarea[id="additional_information"]')
       expect(additionalInformationTextArea.exists()).toEqual(true)
     })
 
+    it('renders the additional information label with required', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/views/ScreeningDecisionFormSpec.jsx]: Need another test for when field is not required.

File: app/javascript/containers/screenings/DecisionFormContainer.jsx [hunk 9]
```diff
@@ -73,6 +76,9 @@ const mergeProps = (stateProps, dispatchProps) => {
     if (field === 'access_restrictions' && value === '') {
       dispatch(setField({field: 'restrictions_rationale', value: null}))
     }
+    if (field === 'additional_information' && value === '') {
```
Comment (saraOSI) [inline app/javascript/containers/screenings/DecisionFormContainer.jsx]: Is this functionality we want to implement? I don't see it mentioned in the story that we should clear the additional information field if the user changes the screening decision. 

If we do need to implement it, is there a feature spec for this?

File: app/javascript/selectors/screening/decisionFormSelectors.js [hunk 10]
```diff
@@ -39,7 +39,8 @@ export const getErrorsSelector = createSelector(
   getDecisionDetailValueSelector,
   (state) => state.getIn(['screeningDecisionForm', 'restrictions_rationale', 'value']) || '',
   (state) => state.get('allegationsForm', List()),
-  (decision, decisionDetail, restrictionsRationale, allegations) => fromJS({
+  (state) => state.getIn(['screeningDecisionForm', 'additional_information', 'value']) || '',
```
[the remark this answers is not in the document]
Comment (saraOSI) [reply app/javascript/selectors/screening/decisionFormSelectors.js]: Agreed with Aman. Code Climate is complaining about the complexity here.

File: app/javascript/selectors/screening/decisionFormSelectors.js [hunk 11]
```diff
@@ -103,9 +109,16 @@ export const getDecisionSelector = createSelector(
   (value, errors) => Map({value, errors})
 )
 
+export const getAdditionalInfoRequiredSelector = createSelector(
+  getDecisionValueSelector,
+  getDecisionDetailValueSelector,
+  (decision, decisionDetail) => (decision && decisionDetail && decision === 'screen_out' && decisionDetail === 'evaluate_out')
```
Comment (saraOSI) [inline app/javascript/selectors/screening/decisionFormSelectors.js]: More of style issue, but it's nice to group the equality and existence checks for each individual parameter. eg:
```
decision && decision === 'screen_out' && decisionDetail && decisionDetail === 'evaluate_out'
```

File: app/javascript/selectors/screening/decisionShowSelectors.js [hunk 12]
```diff
@@ -33,13 +34,28 @@ export const getErrorsSelector = createSelector(
...
+      ),
       restrictions_rationale: combineCompact(
         isRequiredIfCreate(restrictionsRationale, 'Please enter an access restriction reason', () => (accessRestrictions))
       ),
     })
   )
 )
 
+export const getAdditionalInfoRequiredSelector = createSelector(
+  getScreeningSelector,
```
Comment (saraOSI) [inline app/javascript/selectors/screening/decisionShowSelectors.js]: Since you're using createSelector here, it would be better to replace the `const`'s from inside the function with parameters passed to `createSelector`. eg:

```
createSelector(
(state) => state.getIn(['screening', 'screening_decision']),
(state) => state.getIn(['screening', 'screening_decision_detail']),
(decision, decisionDetail) => # check for validity here using decision and decisionDetail
```

File: app/javascript/views/ScreeningDecisionForm.jsx [hunk 13]
```diff
@@ -156,4 +166,4 @@ ScreeningDecisionForm.propTypes = {
   sdmPath: PropTypes.string,
 }
 
-export default ScreeningDecisionForm
+export default ScreeningDecisionForm
```
Comment (saraOSI) [inline app/javascript/views/ScreeningDecisionForm.jsx]: Looks like there is a new trailing line break on the end of the line. Can you run lint and make sure it isn't complaining?

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 14]
```diff
@@ -7,10 +7,14 @@
   let(:error_message) { 'Please enter at least one allegation to promote to referral.' }
   let(:perpetrator) { FactoryGirl.create(:participant, :perpetrator) }
   let(:victim) { FactoryGirl.create(:participant, :victim) }
+  let(:screening_decision_detail) { nil }
```
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: Is there a reason we want to declare default values for `screening_decision` and `screening_decision_detail`? Shouldn't each of the individual contexts define their own?

File: spec/features/screening/validations/screening_decision_validations_spec.rb [hunk 15]
```diff
@@ -136,6 +140,35 @@
...
+      end
+
+      scenario 'card displays errors until user enters additional information' do
+        validate_message_as_user_interacts_with_card(
+          invalid_screening: screening,
+          card_name: 'decision',
+          error_message: error_message,
+          screening_updates: { additional_information: 'My reason for evaluating out' }
+        ) do
+          within '#decision-card.edit' do
```
Comment (saraOSI) [inline spec/features/screening/validations/screening_decision_validations_spec.rb]: A better within block for this (which doesn't depend on the css id) is: `within '.card', text: 'Decision'`

File: spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js [hunk 16]
```diff
@@ -126,6 +127,26 @@ describe('screeningDecisionFormSelectors', () => {
...
   })
 
+  describe('getAdditionalInfoRequiredSelector', () => {
+    it('returns the true if screening decision is screen_out and screening_decision_detail is evaluate_out', () => {
+      const screeningDecisionForm = {screening_decision: {value: 'screen_out'}, screening_decision_detail: {value: 'evaluate_out'}}
+      const state = fromJS({screeningDecisionForm})
+      expect(getAdditionalInfoRequiredSelector(state)).toEqualImmutable(fromJS(true))
+    })
+
+    it('returns the false if screening decision is screen_out and screening_decision_detail is not evaluate_out', () => {
```
Comment (saraOSI) [inline spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js]: s/returns the false/returns false

File: app/javascript/selectors/screening/decisionFormSelectors.js [hunk 17]
```diff
@@ -103,9 +109,16 @@ export const getDecisionSelector = createSelector(
...
+)
+
 export const getAdditionalInformationSelector = createSelector(
   (state) => state.getIn(['screeningDecisionForm', 'additional_information', 'value']),
-  (value) => Map({value: value || ''})
+  (state) => getVisibleErrorsSelector(state).get('additional_information'),
+  (value, errors) => Map({value: value || '', errors})
```
[the remark this answers is not in the document]
Comment (saraOSI) [reply app/javascript/selectors/screening/decisionFormSelectors.js]: 👍  to Aman's comment. Where's the test for errors being a part of this payload?

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js [hunk 18]
```diff
@@ -183,4 +228,24 @@ describe('allegationShowSelectors', () => {
       expect(getRestrictionRationaleSelector(state).get('value')).toEqual('')
     })
   })
+
+  describe('getAdditionalInformationSelector', () => {
+    it('returns errors for screening decision', () => {
```
Comment (saraOSI) [inline spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: Maybe add a test for returning an empty list if no errors are present?

File: spec/javascripts/views/ScreeningDecisionFormSpec.jsx [hunk 19]
```diff
@@ -273,12 +275,24 @@ describe('ScreeningDecisionForm', () => {
...
+      const component = renderScreeningDecisionForm({isAdditionalInfoRequired: true})
+      const additionalInformationLabel = component.find('FormField[htmlFor="additional_information"]')
+      expect(additionalInformationLabel.props().required).toEqual(true)
+    })
+
+    it('renders the additional information label without required', () => {
```
Comment (saraOSI) [inline spec/javascripts/views/ScreeningDecisionFormSpec.jsx]: I don't see a test for onBlur being called if you blur the additional information field.

File: app/javascript/selectors/screening/decisionFormSelectors.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply app/javascript/selectors/screening/decisionFormSelectors.js]: covered this in recent commits.

File: spec/javascripts/views/ScreeningDecisionFormSpec.jsx (hunk 8 above)
Comment (NagaPradeepGajula-osi) [reply spec/javascripts/views/ScreeningDecisionFormSpec.jsx]: done in recent commits.

File: app/javascript/containers/screenings/DecisionFormContainer.jsx (hunk 9 above)
Comment (NagaPradeepGajula-osi) [reply app/javascript/containers/screenings/DecisionFormContainer.jsx]: I have gone through the story again and we don't need this. So, I will be removing it. Thanks for pointing out.

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js (hunk 18 above)
Comment (NagaPradeepGajula-osi) [reply spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: I haven't seen any tests for restriction rational so skipped it.

File: app/javascript/views/ScreeningDecisionForm.jsx (hunk 13 above)
Comment (NagaPradeepGajula-osi) [reply app/javascript/views/ScreeningDecisionForm.jsx]: no offenses reported in lint. all 193 are success.

File: app/javascript/selectors/screening/decisionShowSelectors.js [hunk 20]
```diff
@@ -11,7 +11,8 @@ export const getErrorsSelector = createSelector(
   (state) => state.getIn(['screening', 'access_restrictions']) || '',
   (state) => state.getIn(['screening', 'restrictions_rationale']) || '',
   (state) => state.get('allegationsForm', List()),
-  (decision, decisionDetail, accessRestrictions, restrictionsRationale, allegations) => (
+  (state) => state.getIn(['screening', 'additional_information']) || '',
+  (decision, decisionDetail, accessRestrictions, restrictionsRationale, allegations, additionalInformation) => (
```
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply app/javascript/selectors/screening/decisionShowSelectors.js]: screen shot 2017-12-20 at 10 00 26 am
Code climate says everything is good on github here is the screenshot that I pulled just now.

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js [hunk 21]
```diff
@@ -153,6 +155,29 @@ describe('allegationShowSelectors', () => {
       })
     })
 
+    describe('additional information', () => {
```
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: So the story I am working on 409 has the scope handling the scenario when the decision is screen_out and decision detail is evaluate_out. So I was just covering these. Anyhow I added the test case for when decision is not a screen out?

File: spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js [hunk 22]
```diff
@@ -303,6 +324,35 @@ describe('screeningDecisionFormSelectors', () => {
       })
     })
 
+    describe('additional information', () => {
```
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js]: So the story I am working on 409 has the scope handling the scenario when the decision is screen_out and decision detail is evaluate_out. So I was just covering these. Anyhow I added the test case for when decision is not a screen out?

File: spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js [hunk 23]
```diff
@@ -303,6 +324,42 @@ describe('screeningDecisionFormSelectors', () => {
...
+      })
+
+      it('does not include an error message if decision is screen out, and decision detail is empty', () => {
+        const screeningDecisionForm = {screening_decision: {value: 'screen_out'}}
+        const state = fromJS({screeningDecisionForm})
+        expect(getErrorsSelector(state).get('additional_information'))
+          .toEqualImmutable(List())
+      })
+
+      it('does not include an error message if decision is not screen out', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/selectors/screening/decisionFormSelectorsSpec.js]: 👍

File: spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js [hunk 24]
```diff
@@ -153,6 +155,36 @@ describe('allegationShowSelectors', () => {
...
+      })
+
+      it('does not include an error message if decision is screen out, and decision detail is empty', () => {
+        const screening = {screening_decision: 'screen_out'}
+        const state = fromJS({screening})
+        expect(getErrorsSelector(state).get('additional_information'))
+          .toEqualImmutable(List())
+      })
+
+      it('does not include an error message if decision is not screen out', () => {
```
Comment (AmanShurafa) [inline spec/javascripts/selectors/screening/decisionShowSelectorsSpec.js]: 👍

File: app/javascript/selectors/screening/decisionShowSelectors.js [hunk 25]
```diff
@@ -35,7 +35,7 @@ export const getErrorsSelector = createSelector(
         }
       ),
       additional_information: combineCompact(
-        isRequiredIfCreate(additionalInformation, 'Please enter Additional Information', () => (
+        isRequiredIfCreate(additionalInformation, 'Please enter additional information', () => (
```
Comment (AmanShurafa) [inline app/javascript/selectors/screening/decisionShowSelectors.js]: 👍 This was the only crucial part
